Rewrite integration test in pytest

martinpitt opened the pull request:
Move it from tasks/ into test/ as it involves more than the tasks container, and could e.g. also grow integration tests for metrics.

Also set up ruff and mypy static code checks, and add a pyproject.toml configuration for these.

-----

 - [x] Fix [mysterious S3 404](https://github.com/cockpit-project/cockpituous/actions/runs/8324760479/job/22777027362?pr=606) (works with local run)
 - [x] Bring back "sit" behaviour on failures
 - [x] Use locally built container if it changed in a PR: [demo run](https://github.com/cockpit-project/cockpituous/actions/runs/8327408760/job/22785017929?pr=606) on commit 027dd89399f88 (with touching container)
 - [x] File a bug for S3 host port (not a blocker, but inconvenient): reported to https://github.com/dcermak/pytest_container/issues/190
 - [x] Fix for `pytest -nauto` (with `sudo dnf install python3-pytest-xdist`)

martinpitt commented:
@allisonkarlitskaya This is ready for review now -- there are still a few hacks for pytest_container issues, but upstream is very nice and responsive, so over time we can hopefully clean them up.

allisonkarlitskaya reviewed (commented):
One very quick comment: please add a separate commit to include all of the test dependencies in the tasks container.

allisonkarlitskaya commented:
When running this in a toolbox I get this pretty quickly:

```
ERROR test/test_deployment.py::test_podman - ValueError: Selected runtime podman does not exist on the system
```

Even though I have a (script wrapper) `podman` in my path.

This would be a fairly serious regression...

martinpitt commented:
@allisonkarlitskaya I tried to make the test parallel friendly. Before, running two `pytest` simultaneously immediately failed. But even now the tests fail, but much more subtly:
```
Error: reading CIDFile: open /tmp/tmplgrkkdaq/cidfile: no such file or directory
RuntimeError: Failed to create container
```
Hence I've kept it as a separate commit. Depending on how important parallelism is to you, I can squash or drop it. I don't want to do parallel runs in the GH workflow, it makes test output much harder to read, and it's fast enough in serial mode.

martinpitt commented:
@allisonkarlitskaya I fixed the test to be parallel friendly. I can now run two `pytest` simultaneously.

martinpitt commented:
@allisonkarlitskaya I added a separate FIXUP to use pytest_container from main. That makes things much nicer. If you object to that, I can drop it, otherwise squash.

martinpitt commented:
@allisonkarlitskaya I [fixed](https://github.com/cockpit-project/cockpituous/compare/1c5a3896d7a0d20d082810602d3c319e0ccb3f53..4148c637da460e649484e3829cb24e5b660cc7f4) parallel runs -- restoring the podman socket permissions on teardown isn't compatible with parallel tests.

martinpitt commented:
In installed python3-pytest-xdist and `pytest -nauto` works now, with some [timeout adjustments](https://github.com/cockpit-project/cockpituous/compare/4148c637da460e649484e3829cb24e5b660cc7f4..0d0878c83de2250dfceb5f0a5709ee83ef4a722e)

martinpitt commented:
> One very quick comment: please add a separate commit to include all of the test dependencies in the tasks container.

Obsolete now -- all of them (pytest, ruff, mypy) are already in tasks.

allisonkarlitskaya commented on test/test_deployment.py:
```diff
@@ -0,0 +1,682 @@
+def mock_git_push(pod: PodData) -> Generator[None, None, None]:
+    """Install tasks/mock-git-push"""
+
+    subprocess.run(['podman', 'cp', str(ROOT_DIR / 'tasks/mock-git-push'), f'{pod.tasks}:/usr/local/bin/git'],
```
I wonder if we can just include this in the pod setup?  Is there ever a time we want to push "for real"?

The interaction of one fixture implicitly modifying the state of another fixture (and then returning None for itself) seems a bit magic to me.

martinpitt replied:
Right, moved to the setup.

allisonkarlitskaya commented on test/test_deployment.py:
```diff
@@ -0,0 +1,682 @@
+    generate_config(config,
+                    forge_opts=f'api-url = "{GHAPI_URL_POD}"',
+                    run_args=f'"--pod={pod.pod}", "--env=GITHUB_API={GHAPI_URL_POD}"')
+    return None  # silence ruff PT004
```
I'd much rather these returned a path object which we could pass to `job-runner -F`...

martinpitt replied:
But then we'd diverge from what happens in the actual deployment, where we don't use `-F`, and instead point to it via env variable. I can return the path of course, but it's not actually going to use it. It looks a bit nicer, though.

allisonkarlitskaya reviewed (commented):
This looks awesome, both compared to the shell version, and the iterations that came before.

For me, the most uncomfortable part about your usage of pytest is how you have a bunch of fixtures which cause side-effects in their environment, but don't return anything.  You then use those fixtures from your test functions (while ignoring their values).

I've given a few comments about ways in which I'd change that.

martinpitt commented on .github/workflows/tests.yml:
```diff
@@ -10,9 +10,11 @@ jobs:
-          sudo apt-get install -y make python3-pyflakes python3-pycodestyle
+          sudo apt-get install -y make python3-pyflakes python3-pycodestyle python3-pip python3-pytest
+          # `pip install .[test]` does not work properly on Ubuntu 22.04
+          sudo pip install ruff mypy types-PyYAML
```
Right, it's the same code as the actual deployment, and dissecting a k8s resource isn't *that* trivial with just string functions. I'm ok with ignoring yaml for mypy, but it seemed nice to cover, as that types lib exists.

martinpitt commented on Makefile:
```diff
@@ -6,6 +6,8 @@ all:
 check:
 	python3 -m pyflakes tasks tasks/container/webhook
 	python3 -m pycodestyle --max-line-length=120 --ignore=E722 tasks tasks/container/webhook
+	if command -v mypy >/dev/null && pip show types-PyYAML >/dev/null; then mypy test; else echo "SKIP: mypy or types-PyYAML not installed"; fi
```
Yeah, that's something I'd like to do as a follow-up, or prerequisite. It involves a tasks container refresh to add it there, as the integration test runs itself inside of the tasks container, and would fail without it. I sent #609, if/when that lands I can drop the `pip show` part.

martinpitt replied:
OK, `# type: ignore[import-untyped]` DTRT, and still typechecks if the types are actually installed. So I went with that.

martinpitt commented on tasks/README.md:
```diff
@@ -60,29 +60,30 @@ Some helpful commands:
 You can get an interactive shell with
 
-    tasks/run-local.sh -i
+    pytest -sm shell
```
`@pytest.mark` is the only way that I found how to properly decorate a test to skip itself. You don't have access to the `pytestconfig` fixture (which is a runtime concept) in a decorator. I suppose I can convert it from a decorator into a procedural call though. But then I'd have to add this check to *all* other tests (to skip themselves in `--shell` mode), which I find super ugly.

So in essence, I need an elegant way to say "discover/run these tests by default", but with some special invocation only run `test_shell`; `-k` only gives me half of that.

If you have a better idea, I'm all ears! I don't like the `"not shell"` hack either, but I don't know a better way.

allisonkarlitskaya replied:
You can implement commandline argument parsing via plugins and you can modify test collection via plugins.  I feel like that's all you need?

martinpitt replied:
Ah, thanks for the hint! That goes to the pilot board as improvement then. I'll probably look at it in the next days, I want to learn this.

martinpitt commented on test/test_deployment.py:
```diff
@@ -0,0 +1,682 @@
+
+
+@pytest.fixture(scope='session')
+def pod(config: Config, pytestconfig) -> Generator[PodData, None, None]:
```
Oh, that's nicer indeed! I was following mypy's recommendation without thinking about it too much. Changed to `Iterator`.

martinpitt commented on test/test_deployment.py:
```diff
@@ -0,0 +1,682 @@
+
+    yield f'export GITHUB_API={GHAPI_URL_POD}; SHA={bots_sha}'
+
+    exec_c_out(pod.tasks, f'while kill {mock_pid}; do sleep 0.1; done')
```
We can, but then it doesn't fit nicely into the `exec_c{,_out}()` helpers any more, with `subprocess.run()`, plus it makes the whole `podman exec` stick around. But I rewrote it using `Popen()` now, and it does look quite a bit nicer.

martinpitt commented:
@allisonkarlitskaya I addressed most of your comments, I just don't have a better idea about the `-sm shell`, see the [discussion above](https://github.com/cockpit-project/cockpituous/pull/606#discussion_r1533867604).

allisonkarlitskaya reviewed (approved):
I'm happy enough with this now.  We can address remaining issues/tweaks as follow-ups :)

This is a very nice improvement.  I'm glad you were inspired to work on it :sparkles:

martinpitt commented:
Ack! Let's see how this works in practice. I'll work on a corresponding bots change to use the new test there also, otherwise we break PRs. I added the "shell hack" thing to the pilot board.
